extensions/artillerytables/artillerytables.cpp

```diff
@@ -0,0 +1,356 @@
+            return { -1, -1, -1 };
+        }
+        radicand = sqrt(radicand);
+        double angleRoot = atan((pow(_muzzleVelocity, 2) + radicand) / (gravityABS * _rangeToHit));
+        if ((angleRoot > _maxElev) || (angleRoot < _minElev)) {
+            angleRoot = atan((pow(_muzzleVelocity, 2) - radicand) / (gravityABS * _rangeToHit));
+        }
+        if ((angleRoot > _maxElev) || (angleRoot < _minElev)) {
+            return { -1, -1, -1 };
+        };
+        const double tof = _rangeToHit / (_muzzleVelocity * cos(angleRoot));
+        return { _rangeToHit, angleRoot, tof };
+    }
+
+
+    int numberOfAttempts = 0;
+    double resultDistance = -1;
+    double resultTime = -1;
+    double currentError = 9999;
+    double currentElevation = -1;
+    do {
+        if (numberOfAttempts++ > 50) break; // for safetey, min/max should converge long before
+        currentElevation = (searchMin + searchMax) / 2.0;
+        std::tie(std::ignore, resultDistance, resultTime) = simulateShot(currentElevation, _muzzleVelocity, _heightToHit, 0, 0, 15, 1, _airFriction);
+        currentError = _rangeToHit - resultDistance;
+        // printf("elev %f [%f, %f]range%f\n goes %f [%f]\n", currentElevation, searchMin, searchMax, (searchMax - searchMin), resultDistance, currentError);
+        if ((currentError > 0) ^ (!_highArc)) {
+            searchMax = currentElevation;
+        } else {
+            searchMin = currentElevation;
+        }
+    } while ((searchMax - searchMin) > rangeSearchAngleConvergance);
+
+    // printf("[%f, %f] Actuall [%f] err [%f of %f]\n", _rangeToHit, _heightToHit, resultDistance, currentError, (_rangeToHit * rangeSearchErrorMax * (_highArc ? 1.0 : 2.0)));
+    // On some low angle shots, it will really struggle to converge because of precision issues
+    if ((abs(currentError) > (_rangeToHit * rangeSearchErrorMax * (_highArc ? 1.0 : 2.0)))) {
+        return { -1, -1, -1 };
+    }
+    return { resultDistance, currentElevation, resultTime };
+}
+
+void writeNumber(std::stringstream & ss, double _num, const int& _widthInt, const int& _widthDec) {
+    if ((_num < 0) && (_num > -0.05)) { // hard coded fix -0.0 rounding errors
+        _num = 0;
+    }
+    if (_widthInt > 1) {
+        ss << std::setfill('0');
+    }
+    ss << std::fixed << std::setw(_widthInt) << std::setprecision(_widthDec) << _num;
+}
+
+std::string simulateCalcRangeTableLine(const double& _rangeToHit, const double& _muzzleVelocity, const double& _airFriction, const double& _minElev, const double& _maxElev, const bool& _highArc) {
+    auto [actualDistance, lineElevation, lineTimeOfFlight] = simulateFindSolution(_rangeToHit, 0, _muzzleVelocity, _airFriction, _minElev, _maxElev, _highArc);
+    if (lineTimeOfFlight < 0) {
+        return "";
+    }
+    auto [actualDistanceHeight, lineHeightElevation, lineHeightTimeOfFlight] = simulateFindSolution(_rangeToHit, -100, _muzzleVelocity, _airFriction, _minElev, _maxElev, _highArc);
```

Comment:
Wasn't there some reason in the past why we didn't want to use C++17?

Reply:
Yes, reason was it just came out and compiler support was not ideal. I still don't see a reason to use it for one line though.

Reply:
gcc support for C++17 is still abysmal. And we compile our linux plugins with gcc.
Although this plugin should be clientside only, so windows only. So it is kinda fine.
I didn't check if there is more C++17 in here.

Reply:
but it's 2019 :smile: 
I think this was the only 17 feature, should compile anywhere now